addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/Powerline546EHandler.java

```diff
@@ -0,0 +1,153 @@
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.types.Command;
+import org.openhab.binding.avmfritz.BindingConstants;
+import org.openhab.binding.avmfritz.internal.ahamodel.AVMFritzBaseModel;
+import org.openhab.binding.avmfritz.internal.ahamodel.SwitchModel;
+import org.openhab.binding.avmfritz.internal.config.AVMFritzConfiguration;
+import org.openhab.binding.avmfritz.internal.hardware.FritzahaWebInterface;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handler for a FRITZ!Powerline 546E device. Handles polling of values from AHA devices and commands, which are sent to
+ * one of the channels.
+ *
+ * @author Robert Bausdorf - Initial contribution
+ * @author Christoph Weitkamp - Added support for groups
+ *
+ */
+public class Powerline546EHandler extends AVMFritzBaseBridgeHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(Powerline546EHandler.class);
+
+    /**
+     * keeps track of the current state for handling of increase/decrease
+     */
+    private AVMFritzBaseModel state;
+
+    /**
+     * Constructor
+     *
+     * @param bridge Bridge object representing a FRITZ!Powerline 546E
+     */
+    public Powerline546EHandler(@NonNull Bridge bridge) {
+        super(bridge);
+    }
+
+    @Override
+    public void addDeviceList(AVMFritzBaseModel device) {
+        logger.debug("set device model: {}", device);
+        ThingUID thingUID = getThingUID(device);
+        if (getThing().getUID().equals(thingUID)) {
+            logger.debug("update self {} with device model: {}", thingUID, device);
+            setState(device);
+            updateThingFromDevice(getThing(), device);
+        } else {
+            super.addDeviceList(device);
+        }
+    }
+
+    /**
+     * Updates things from device model.
+     *
+     * @param thing Thing to be updated.
+     * @param device Device model with new data.
+     */
+    @Override
+    protected void updateThingFromDevice(Thing thing, AVMFritzBaseModel device) {
+        // save AIN to config for FRITZ!Powerline 546E stand-alone
+        if (thing.getConfiguration().get(THING_AIN) == null) {
+            thing.getConfiguration().put(THING_AIN, device.getIdentifier());
+        }
+        super.updateThingFromDevice(thing, device);
+    }
+
+    /**
+     * Builds a {@link ThingUID} from a device model. The UID is build from the
+     * {@link BindingConstants#BINDING_ID} and value of
+     * {@link AVMFritzBaseModel#getProductName()} in which all characters NOT matching
+     * the regex [^a-zA-Z0-9_] are replaced by "_".
+     *
+     * @param device Discovered device model
+     * @return ThingUID without illegal characters.
+     */
+    @Override
+    public ThingUID getThingUID(AVMFritzBaseModel device) {
+        ThingTypeUID thingTypeUID = new ThingTypeUID(BINDING_ID,
```

Comment:
It appears that a part of this class is copy for example `.replaceAll("[^a-zA-Z0-9_]", "_")` seems to be duplicated

Reply:
14x so maybe time to put on another level, maybe a sanitization method on the level of ThingTypeUID would make sense `ThingTypeUID.sanitizeThingTypeId`